eureka2-core/src/main/java/com/netflix/eureka2/transport/base/BaseMessageConnection.java

```diff
@@ -267,29 +295,73 @@ public void call() {
                 .doOnCompleted(new Action0() {
                     @Override
                     public void call() {
-                        pendingAck.add(ack);
+                        pendingAckQueue.add(ack);
+                        metrics.incrementPendingAckCounter();
                         metrics.incrementOutgoingMessageCounter(message.getClass(), 1);
+
+                        // Connection might be closed when we serve this request, and since
+                        // pendingAckQueue and closed variables are not changed together atomically, we check
+                        // close status after adding item to pendingAckQueue, and optionally drain it.
+                        if (closed.get()) {
+                            drainPendingAckQueue();
```

Comment:
Is this threadsafe? Seems like this can be called by multiple observable chains concurrently here.